File: src/policy.ts
import * as checks from "./checks.js";
import { presets } from "./presets.js";
import type { Check, GateOptions, Policy, PolicyEntry, Priority } from "./types.js";

type Options = Record<string, unknown>;
type Factory = (options: Options) => Check;

const num = (o: Options, key: string): number | undefined => (typeof o[key] === "number" ? (o[key] as number) : undefined);
const str = (o: Options, key: string): string | undefined => (typeof o[key] === "string" ? (o[key] as string) : undefined);
const bool = (o: Options, key: string): boolean | undefined => (typeof o[key] === "boolean" ? (o[key] as boolean) : undefined);
const prio = (o: Options, key: string): Priority | undefined => str(o, key) as Priority | undefined;
const strs = (o: Options, key: string): string[] | undefined => (Array.isArray(o[key]) ? (o[key] as string[]) : undefined);
const opt = <T>(value: T | undefined, key: string): Record<string, T> => (value === undefined ? {} : { [key]: value });

const budgetOptions = (o: Options) => ({ ...opt(num(o, "limit"), "limit"), ...opt(prio(o, "bypassPriority"), "bypassPriority"), ...opt(num(o, "nearLimit"), "nearLimit") });

/** Every check a JSON policy may name, with the options it reads. */
export const KNOWN_CHECKS: Record<string, Factory> = {
  killSwitch: (o) => checks.killSwitch(() => bool(o, "on") === true),
  consent: () => checks.consent(),
  enabled: () => checks.enabled(),
  mode: (o) => checks.mode({ allow: strs(o, "allow") ?? ["normal"] }),
  snooze: (o) => checks.snooze({ ...opt(bool(o, "defer"), "defer") }),
  mute: () => checks.mute(),
  intensity: (o) => (o.floors ? checks.intensity(o.floors as Record<"low" | "normal" | "high", Priority>) : checks.intensity()),
  quietHours: (o) => checks.quietHours({ ...opt(prio(o, "priorityFloor"), "priorityFloor") }),
  trustRamp: (o) => checks.trustRamp({ ...opt(num(o, "days"), "days"), ...opt(prio(o, "minPriority"), "minPriority") }),
  dismissalCooldown: (o) => checks.dismissalCooldown({ ...opt(num(o, "dismissals"), "dismissals"), ...opt(num(o, "withinDays"), "withinDays"), ...opt(num(o, "silenceDays"), "silenceDays") }),
  adaptiveTiming: () => checks.adaptiveTiming(),
  dailyBudget: (o) => checks.dailyBudget(budgetOptions(o)),
  weeklyBudget: (o) => checks.weeklyBudget(budgetOptions(o)),
  monthlyBudget: (o) => checks.monthlyBudget(budgetOptions(o)),
  utilityFloor: (o) => checks.utilityFloor({ costFalseAlarm: num(o, "costFalseAlarm") ?? 1, costMissedHelp: num(o, "costMissedHelp") ?? 1 }),
  boundedDeferral: (o) => checks.boundedDeferral({ ...opt(num(o, "lambda"), "lambda"), ...opt(num(o, "interruptCost"), "interruptCost"), ...opt(num(o, "staleness"), "staleness"), ...opt(num(o, "boundSeconds"), "boundSeconds") }),
  allowedWindow: (o) => checks.allowedWindow({ start: str(o, "start") ?? "08:00", end: str(o, "end") ?? "21:00", timezone: str(o, "timezone") ?? "user", ...opt(prio(o, "priorityFloor"), "priorityFloor"), ...opt(str(o, "id"), "id") }),
  requiresConsent: (o) => checks.requiresConsent({ name: str(o, "name") ?? "consent", ...(o.when ? { when: o.when as { start: string; end: string; timezone: string } } : {}), ...opt(str(o, "id"), "id") }),
  rateLimit: (o) => checks.rateLimit({ limit: num(o, "limit") ?? 1, perSeconds: num(o, "perSeconds") ?? 1, ...opt(str(o, "keyBy") as "user" | "channel" | undefined, "keyBy"), ...opt(str(o, "id"), "id") }),
  recentInteraction: (o) => checks.recentInteraction({ withinHours: num(o, "withinHours") ?? 48 }),
  windowBudget: (o) => checks.windowBudget({ limit: num(o, "limit") ?? 1, withinHours: num(o, "withinHours") ?? 48 }),
};

const SUPPORTED_MAJOR = 1;

/** Compile a JSON policy into gate options. Throws on unknown ids, unknown presets, or an unsupported specVersion. */
export function compilePolicy(policy: Policy): GateOptions {
  const major = Number(String(policy.specVersion).split(".")[0]);
  if (!Number.isInteger(major) || major !== SUPPORTED_MAJOR) {
    throw new Error(`policy specVersion ${policy.specVersion} is not supported; this package implements spec ${SUPPORTED_MAJOR}.x`);
  }
  if (!Array.isArray(policy.checks) || !policy.checks.length) throw new Error("policy.checks must be a non-empty array");
  const compiled: Check[] = [];
  for (const entry of policy.checks as PolicyEntry[]) {
    const { shadow, ...rest } = entry as Options & { shadow?: boolean };
    let built: Check[];
    if (typeof rest.preset === "string") {
      const preset = presets[rest.preset];
      if (!preset) throw new Error(`unknown preset "${rest.preset}"; known presets: ${Object.keys(presets).join(", ")}`);
      const { preset: _name, ...options } = rest;
      built = preset(options);
    } else if (typeof rest.id === "string") {
      const factory = KNOWN_CHECKS[rest.id];
      if (!factory) throw new Error(`unknown check "${rest.id}"; known checks: ${Object.keys(KNOWN_CHECKS).join(", ")}`);
      const { id: _id, ...options } = rest;
      built = [factory(options)];
    } else {
      throw new Error("each policy entry needs an id or a preset");
    }
    if (shadow) for (const c of built) c.shadow = true;
    compiled.push(...built);
  }
  return {
    checks: compiled,
    ...(policy.onStoreError ? { onStoreError: policy.onStoreError } : {}),
    ...(policy.keyPrefix !== undefined ? { keyPrefix: policy.keyPrefix } : {}),
  };
}

File: go/proactivegate/policy.go
package proactivegate

import (
	"fmt"
	"sort"
	"strconv"
	"strings"
)

// Policy is a JSON policy document (spec/schema/policy.schema.json): the same
// checks as data.
type Policy struct {
	SpecVersion  string        `json:"specVersion"`
	OnStoreError string        `json:"onStoreError"`
	KeyPrefix    string        `json:"keyPrefix"`
	Checks       []PolicyEntry `json:"checks"`
}

// PolicyEntry is one entry of the checks array: either { "id": <check> } or
// { "preset": <name> }, each with options and an optional "shadow" flag.
type PolicyEntry map[string]any

const supportedMajor = 1

type options map[string]any

func (o options) num(key string) float64 {
	if v, ok := o[key].(float64); ok {
		return v
	}
	return 0
}

func (o options) str(key string) string {
	if v, ok := o[key].(string); ok {
		return v
	}
	return ""
}

func (o options) bool(key string) bool {
	v, _ := o[key].(bool)
	return v
}

func (o options) strs(key string) []string {
	list, ok := o[key].([]any)
	if !ok {
		return nil
	}
	out := make([]string, 0, len(list))
	for _, v := range list {
		if s, ok := v.(string); ok {
			out = append(out, s)
		}
	}
	return out
}

func (o options) prio(key string) Priority {
	if v, ok := o[key].(string); ok {
		return Priority(v)
	}
	return ""
}

func (o options) budgetOptions() budgetOptions {
	return budgetOptions{
		Limit:          int64(o.num("limit")),
		BypassPriority: o.prio("bypassPriority"),
		NearLimit:      o.num("nearLimit"),
	}
}

// knownChecks is every check a JSON policy may name, with the options it reads.
var knownChecks = map[string]func(options) (Check, error){
	"killSwitch": func(o options) (Check, error) { return killSwitch(o.bool("on")), nil },
	"consent":    func(o options) (Check, error) { return consent(), nil },
	"enabled":    func(o options) (Check, error) { return enabled(), nil },
	"mode": func(o options) (Check, error) {
		allow := o.strs("allow")
		if allow == nil {
			allow = []string{"normal"}
		}
		return mode(allow), nil
	},
	"snooze": func(o options) (Check, error) { return snooze(o.bool("defer")), nil },
	"mute":   func(o options) (Check, error) { return mute(), nil },
	"intensity": func(o options) (Check, error) {
		var floors map[string]Priority
		if raw, ok := o["floors"].(map[string]any); ok {
			floors = map[string]Priority{}
			for k, v := range raw {
				if s, ok := v.(string); ok {
					floors[k] = Priority(s)
				}
			}
		}
		return intensity(floors), nil
	},
	"quietHours": func(o options) (Check, error) { return quietHours(o.prio("priorityFloor")), nil },
	"trustRamp": func(o options) (Check, error) {
		return trustRamp(o.num("days"), o.prio("minPriority")), nil
	},
	"dismissalCooldown": func(o options) (Check, error) {
		return dismissalCooldown(int64(o.num("dismissals")), o.num("withinDays"), o.num("silenceDays")), nil
	},
	"adaptiveTiming": func(o options) (Check, error) { return adaptiveTiming(), nil },
	"dedupe":         func(o options) (Check, error) { return dedupe(int64(o.num("windowSeconds"))), nil },
	"dailyBudget":    func(o options) (Check, error) { return dailyBudget(o.budgetOptions()), nil },
	"weeklyBudget":   func(o options) (Check, error) { return weeklyBudget(o.budgetOptions()), nil },
	"monthlyBudget":  func(o options) (Check, error) { return monthlyBudget(o.budgetOptions()), nil },
	"utilityFloor": func(o options) (Check, error) {
		cfa, cfn := o.num("costFalseAlarm"), o.num("costMissedHelp")
		if cfa == 0 {
			cfa = 1
		}
		if cfn == 0 {
			cfn = 1
		}
		return utilityFloor(cfa, cfn), nil
	},
	"boundedDeferral": func(o options) (Check, error) {
		return boundedDeferral(o.num("lambda"), o.num("interruptCost"), o.num("staleness"), o.num("boundSeconds")), nil
	},
	"allowedWindow": func(o options) (Check, error) {
		start, end, timezone := o.str("start"), o.str("end"), o.str("timezone")
		if start == "" {
			start = "08:00"
		}
		if end == "" {
			end = "21:00"
		}
		if timezone == "" {
			timezone = "user"
		}
		return allowedWindow(start, end, timezone, o.prio("priorityFloor"), o.str("id"))
	},
	"requiresConsent": func(o options) (Check, error) {
		name := o.str("name")
		if name == "" {
			name = "consent"
		}
		var when *QuietWindow
		var whenZone string
		if raw, ok := o["when"].(map[string]any); ok {
			wo := options(raw)
			when = &QuietWindow{Start: wo.str("start"), End: wo.str("end")}
			whenZone = wo.str("timezone")
		}
		return requiresConsent(name, when, whenZone, o.str("id"))
	},
	"rateLimit": func(o options) (Check, error) {
		limit, perSeconds := int64(o.num("limit")), int64(o.num("perSeconds"))
		if limit == 0 {
			limit = 1
		}
		if perSeconds == 0 {
			perSeconds = 1
		}
		return rateLimit(limit, perSeconds, o.str("keyBy"), o.str("id")), nil
	},
	"recentInteraction": func(o options) (Check, error) {
		within := o.num("withinHours")
		if within == 0 {
			within = 48
		}
		return recentInteraction(within), nil
	},
	"windowBudget": func(o options) (Check, error) {
		limit, within := int64(o.num("limit")), o.num("withinHours")
		if limit == 0 {
			limit = 1
		}
		if within == 0 {
			within = 48
		}
		return windowBudget(limit, within), nil
	},
}

func sortedKeys[V any](m map[string]V) []string {
	out := make([]string, 0, len(m))
	for k := range m {
		out = append(out, k)
	}
	sort.Strings(out)
	return out
}

// CompilePolicy compiles a JSON policy into gate options. It errors on an
// unsupported specVersion, an unknown check id or an unknown preset name,
// naming the known ids (7.3).
func CompilePolicy(policy *Policy) (GateOptions, error) {
	major, err := strconv.Atoi(strings.SplitN(policy.SpecVersion, ".", 2)[0])
	if err != nil || major != supportedMajor {
		return GateOptions{}, fmt.Errorf("policy specVersion %s is not supported; this package implements spec %d.x", policy.SpecVersion, supportedMajor)
	}
	if len(policy.Checks) == 0 {
		return GateOptions{}, fmt.Errorf("policy.checks must be a non-empty array")
	}
	var compiled []Check
	for _, entry := range policy.Checks {
		shadow, _ := entry["shadow"].(bool)
		rest := options(entry)
		delete(rest, "shadow")
		var built []Check
		if presetName, ok := rest["preset"].(string); ok {
			preset, known := Presets[presetName]
			if !known {
				return GateOptions{}, fmt.Errorf("unknown preset %q; known presets: %s", presetName, strings.Join(sortedKeys(Presets), ", "))
			}
			delete(rest, "preset")
			built, err = preset.Build(rest)
			if err != nil {
				return GateOptions{}, err
			}
		} else if id, ok := rest["id"].(string); ok {
			factory, known := knownChecks[id]
			if !known {
				return GateOptions{}, fmt.Errorf("unknown check %q; known checks: %s", id, strings.Join(sortedKeys(knownChecks), ", "))
			}
			delete(rest, "id")
			c, err := factory(rest)
			if err != nil {
				return GateOptions{}, err
			}
			built = []Check{c}
		} else {
			return GateOptions{}, fmt.Errorf("each policy entry needs an id or a preset")
		}
		if shadow {
			for i := range built {
				built[i].Shadow = true
			}
		}
		compiled = append(compiled, built...)
	}
	return GateOptions{
		Checks:       compiled,
		OnStoreError: policy.OnStoreError,
		KeyPrefix:    policy.KeyPrefix,
	}, nil
}

// NewGateFromPolicy compiles a JSON policy document and binds it to store.
// A nil store means a fresh MemoryStore.
func NewGateFromPolicy(policy *Policy, store Store) (*Gate, error) {
	opts, err := CompilePolicy(policy)
	if err != nil {
		return nil, err
	}
	opts.Store = store
	return NewGate(opts), nil
}
